functions/slack_slash_command/index.php

```diff
@@ -0,0 +1,171 @@
+    $signature = $request->getHeader("X-Slack-Signature");
+    if (!$timestamp || !$signature) {
+        return false;
+    } else {
+        $timestamp = $timestamp[0];
+        $signature = $signature[0];
+    }
+
+    // Compute signature
+    $plaintext = 'v0:' . $timestamp . ':' . (string) $request->getBody();
+    $hash = 'v0=' . hash_hmac('sha256', $plaintext, $SLACK_SECRET);
+
+    return $hash === $signature;
+}
+// [END functions_verify_webhook]
+
+// [START functions_slack_format]
+/**
+* Format the Knowledge Graph API response into a richly formatted Slack message.
+*/
+function formatSlackMessage($kgResponse, $query): string
+{
+    $responseJson = [
+        "response_type" => "in_channel",
+        "text" => "Query: " . $query
+    ];
+
+    $entityList = $kgResponse["itemListElement"];
+
+    // Extract the first entity from the result list, if any
+    if (empty($entityList)) {
+        $attachmentJson = ["text" => "No results match your query..."];
+        $responseJson["attachments"] = $attachmentJson;
+
+        return json_encode($responseJson);
+    }
+
+    $entity = $entityList[0]["result"];
+
+    // Construct Knowledge Graph response attachment
+    $title = $entity["name"];
+    if (isset($entity["description"])) {
+        $title = $title . " " . $entity["description"];
+    }
+    $attachmentJson = ["title" => $title];
+
+    if (isset($entity["detailedDescription"])) {
+        $detailedDescJson = $entity["detailedDescription"];
+        $attachmentJson = array_merge([
+            "title_link" => $detailedDescJson[ "url"],
+            "text" => $detailedDescJson["articleBody"],
+        ], $attachmentJson);
+    }
+
+    if ($entity["image"]) {
+        $imageJson = $entity["image"];
+        $attachmentJson["image_url"] = $imageJson["contentUrl"];
+    }
+
+    $responseJson["attachments"] = array($attachmentJson);
+
+    return json_encode($responseJson);
+}
+// [END functions_slack_format]
+
+// [START functions_slack_request]
+/**
+* Send the user's search query to the Knowledge Graph API.
+*/
+function searchKnowledgeGraph($query): Google_Service_Kgsearch_SearchResponse
+{
+    /**
+    * PHP Functions Framework does not support "global"/instance-scoped
+    * constants, so we fetch these values within PHP functions themselves.
+    */
+    $API_KEY = getenv("KG_API_KEY");
+
+    $apiClient = new Google\Client();
+    $apiClient->setDeveloperKey($API_KEY);
+
+    $service = new Google_Service_Kgsearch($apiClient);
+
+    $params = ['query' => $query];
+
+    $kgResults = $service->entities->search($params);
+
+    return $kgResults;
+}
+// [END functions_slack_request]
+
+// [START functions_slack_search]
+/**
+* Receive a Slash Command request from Slack.
+*/
+function receiveRequest(ServerRequestInterface $request): ResponseInterface
+{
+    // Validate request
+    if ($request->getMethod() !== 'POST') {
+        // [] = empty headers
+        return new Response(405, [], '');
```

Comment:
I think being explicit with the first parameters was useful for the initial HTTP introductions, but at this point more conciseness makes sense. PHP doesn't require you to specify all the parameters on the caller's side where there's a default in the function signature. Therefore, this could be:

```suggestion
        return new Response(405);
```

However, without any text in the response, this might confuse developers trying out the function to see no response. It would be more helpful to return a response that will show up in non-verbose `curl`, even if only "Method not allowed".

This feedback also applies to the 400 and 403 responses below.

Reply:
Done.